pkg/redislocker/redislocker.go

```diff
@@ -0,0 +1,148 @@
+package redislocker
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"golang.org/x/exp/slog"
+
+	"github.com/tus/tusd/v2/pkg/handler"
+)
+
+var (
+	LockExchangeChannel = "tusd_lock_release_request_%s"
+	LockReleaseChannel  = "tusd_lock_released_%s"
+	LockExpiry          = 8 * time.Second
+)
+
+type LockExchange interface {
+	Listen(ctx context.Context, id string, callback func())
+	Request(ctx context.Context, id string) error
+}
+
+type BidirectionalLockExchange interface {
+	LockExchange
+	Release(ctx context.Context, id string) error
+}
+
+type MutexLock interface {
+	TryLockContext(context.Context) error
+	ExtendContext(context.Context) (bool, error)
+	UnlockContext(context.Context) (bool, error)
+	Until() time.Time
+}
+
+type RedisLocker struct {
+	CreateMutex func(id string) MutexLock
+	Exchange    BidirectionalLockExchange
+	Logger      *slog.Logger
+}
+
+func (locker *RedisLocker) UseIn(composer *handler.StoreComposer) {
+	composer.UseLocker(locker)
+}
+
+func (locker *RedisLocker) NewLock(id string) (handler.Lock, error) {
+	mutex := locker.CreateMutex(id)
+	return &redisLock{
+		id:       id,
+		mutex:    mutex,
+		exchange: locker.Exchange,
+		logger:   locker.Logger.With("upload_id", id),
+	}, nil
+}
+
+type redisLock struct {
+	id       string
+	mutex    MutexLock
+	ctx      context.Context
+	cancel   func()
+	exchange BidirectionalLockExchange
+	logger   *slog.Logger
+}
+
+func (l *redisLock) Lock(ctx context.Context, releaseRequested func()) error {
+	l.logger.Debug("locking upload", "id", l.id)
+	if err := l.requestLock(ctx); err != nil {
+		return err
+	}
+	go l.exchange.Listen(l.ctx, l.id, releaseRequested)
+	go func() {
+		if err := l.keepAlive(l.ctx); err != nil {
```

Comment:
Any error that isn't related to the context ending should be logged here as well. `l.keepAlive` performs some logging, but this could be moved here.